_posts/2020-09-14-notes-on-building-python-for-windows-using-rbm.md

```diff
@@ -0,0 +1,77 @@
+And since the Tor Browser is a secure project, said Python interpreter has to be built reproducibly.
+For this reason, the `tor-browser-build` repository has received some patches to reproducibly build such binaries.
+
+This post details the problems encountered while making them.
+
+Building CPython reproducibly for Linux is trivial.
+Download the source code, patch it to disable build timestamps, and compile.
+That's it, you're done.
+
+The situation is markedly different on Windows.
+The official Python documentation suggests [building with Microsoft Visual Studio 2017](https://github.com/python/cpython/blob/master/PCbuild/readme.txt).
+This is no good.
+That compiler doesn't support reproducible builds, it's closed-source, and it would have to run in Wine.
+
+There should be another way.
+CPython is written in standard C.
+GCC can cross-compile C software for Windows using MinGW - the compiler runs on Linux, but produces binaries for Windows.
+That's how the Tor Project build Tor Browser for Windows.
+In theory, it should be possible to use GCC to cross-compile CPython for Windows, and a lot of Tor's tooling should be possible to re-use.
+
+In practice, Python's ordinary build system doesn't support this.
+It isn't made for cross-compilation.
+Thankfully, one [Erik Janssens](https://github.com/erikjanss) (no relation to the Namecoin project) had created an alternative [Meson](https://mesonbuild.com/)-based [build script](https://github.com/v-finance/cross-python) to compile Python for Windows with Linux using GCC and MinGW.
+Many thanks!
+
+This build script needed some minor adaptations to properly integrate it with the Tor Project's RBM build system.
+Meson and RBM both have built-in features for dependency handling, such as downloading files from URLs, verifying their hashes, and caching them.
+Doing this requires Internet access.
+When compiling, RBM disables Internet access.
+
+The configuration script will thus have to download the binaries and move them into the cache directory used by Meson.
+This change also makes the build faster - RBM uses a fresh VM for each build, so unless the cache is outside of the VM, it will be wiped on each new build.
+
+After this was done, RBM was able to produce something resembling a Python 3.8 interpreter.
+However, it wasn't functional.
+Owing to what presumably was an idiosyncracy of the MinGW system, a critical library named `libwinpthread-1.dll` wasn't included.
+
+This problem turned out to be easy to fix.
+Ruben Van Boxem, a contributor to the MinGW project, [suggested](https://stackoverflow.com/a/14033674) on StackOverflow that the compiler/linker simply be explicitly told to link this library statically.
+This worked:
+
+```
+$ wine python.exe
+Python 3.8.5 (default, xx/xx/xx, xx:xx:xx) [gcc] on win32
+>>>
+```
+
+The changes are now [submitted](https://gitlab.torproject.org/tpo/applications/tor-browser-build/-/merge_requests/66) and subject to review.
```

Comment:
I generally prefer not to link to unmerged PR's, especially not on 3rd-party repos, because it might result in the "peanut gallery" showing up and spamming the ticket, which would irritate the repo maintainers (the Tor devs in this case).

Reply:
Amended. Likewise for Python, I take it.